### PR TITLE
✨ Implement Init plugin subcommand

### PR DESCRIPTION
**Pull request by Fedosin**
**What this PR does / why we need it**:
This PR adds Init subcommand to the operator plugin.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: cmd/plugin/cmd/init.go [hunk 1]
```diff
@@ -67,65 +93,417 @@ var initCmd = &cobra.Command{
...
+	clusterClient := cluster.New(clusterKubeconfig, configClient)
+
+	if err := clusterClient.ProviderComponents().Create(ctx, components.Objs()); err != nil {
+		return fmt.Errorf("cannot create CAPI operator components: %w", err)
+	}
+
+	return nil
+}
+
+func newGenericProvider(providerType clusterctlv1.ProviderType) operatorv1.GenericProvider {
```
Comment (Danil-Grigorev) [inline cmd/plugin/cmd/init.go:407]: We have a similar (but opposite of that) method in utils. Does it make sense to move it there? Also maybe we can leverage `GetType` on the provider wrapper to only store a list of providers

File: cmd/plugin/cmd/init_test.go [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+				g.Expect(err).To(HaveOccurred())
+
+				return
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+
+			for _, genericProvider := range tt.wantedProviders {
+				g.Eventually(func() (bool, error) {
+					provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
```
Comment (Danil-Grigorev) [inline cmd/plugin/cmd/init_test.go:333]: BTW, you can define a custom interface deriving from client.Object and GenericProvider to get rid of the method and reduce it to a single `client.Get`:
```go
type GeneriProviderObject {
    client.Object
    genericprovider.GenericProvider
}
…

if err := client.Get(ctx, types.NamespacedName{Name: genericProvider.GetName(), Namespace: genericProvider.GetNamespace()}, provider); err != nil {
 	return nil, err
}

```

**Review by Danil-Grigorev — COMMENTED** [review]
Looks good, some nits, suggestions

File: cmd/plugin/cmd/init.go [hunk 3]
```diff
@@ -67,65 +93,417 @@ var initCmd = &cobra.Command{
...
+		capioperator init --infrastructure aws --config-secret-name=capa-secret --target-namespace foo`),
 	Args: cobra.NoArgs,
 	RunE: func(cmd *cobra.Command, args []string) error {
 		return runInit()
 	},
 }
 
 func init() {
-	initCmd.PersistentFlags().StringVar(&initOpts.kubeconfig, "kubeconfig", "",
+	initCmd.PersistentFlags().StringVar(&initOpts.kubeconfig, "kubeconfig", filepath.Join(os.Getenv("HOME"), ".kube", "config"),
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply cmd/plugin/cmd/init.go:122]: I checked the utils method, it is not entirely correct, as it does not pick $KUBECONFIG variable, think it is supposed to.

File: cmd/plugin/cmd/init.go [hunk 4]
```diff
@@ -18,27 +18,53 @@ package cmd
...
 
 type initOptions struct {
 	kubeconfig              string
 	kubeconfigContext       string
+	operatorVersion         string
 	coreProvider            string
 	bootstrapProviders      []string
 	controlPlaneProviders   []string
 	infrastructureProviders []string
 	// ipamProviders             []string
```
[the remark this answers is not in the document]
Comment (Fedosin) [reply cmd/plugin/cmd/init.go:53]: yes, I rebased this PR on master.

File: cmd/plugin/cmd/init.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Fedosin) [reply cmd/plugin/cmd/init.go:122]: Thank you for noticing this. Now we check KUBECONFIG env var as well.

File: cmd/plugin/cmd/init.go [hunk 5]
```diff
@@ -67,65 +93,417 @@ var initCmd = &cobra.Command{
...
 	// initCmd.PersistentFlags().StringSliceVar(&initOpts.ipamProviders, "ipam", nil,
 	//	"IPAM providers and versions (e.g. infoblox:v0.0.1) to add to the management cluster.")
 	// initCmd.PersistentFlags().StringSliceVar(&initOpts.runtimeExtensionProviders, "runtime-extension", nil,
 	//	"Runtime extension providers and versions (e.g. test:v0.0.1) to add to the management cluster.")
-	initCmd.PersistentFlags().StringSliceVar(&initOpts.addonProviders, "addon", nil,
+	initCmd.PersistentFlags().StringSliceVar(&initOpts.addonProviders, "addon", []string{},
 		"Add-on providers and versions (e.g. helm:v0.1.0) to add to the management cluster.")
 	initCmd.Flags().StringVarP(&initOpts.targetNamespace, "target-namespace", "n", "capi-operator-system",
 		"The target namespace where the operator should be deployed. If unspecified, the 'capi-operator-system' namespace is used.")
+	initCmd.Flags().StringVar(&initOpts.configSecretName, "config-secret-name", "",
```
[the remark this answers is not in the document]
Comment (Fedosin) [reply cmd/plugin/cmd/init.go:144]: converted these two options into 1: --config-secret, that has <config_secret_name>:<config_secret_namespace> format

File: cmd/plugin/cmd/init.go [hunk 6]
```diff
@@ -67,65 +93,417 @@ var initCmd = &cobra.Command{
...
+		"The config secret name to be used for the management cluster.")
+	initCmd.Flags().StringVar(&initOpts.configSecretNamespace, "config-secret-namespace", "capi-operator-system",
+		"The config secret namespace to be used for the management cluster.")
 	initCmd.Flags().BoolVar(&initOpts.waitProviders, "wait-providers", false,
```
[the remark this answers is not in the document]
Comment (Fedosin) [reply cmd/plugin/cmd/init.go:148]: done

File: cmd/plugin/cmd/init.go (hunk 1 above)
Comment (Fedosin) [reply cmd/plugin/cmd/init.go:407]: Moved this method to utils.go

File: cmd/plugin/cmd/init_test.go (hunk 2 above)
Comment (Fedosin) [reply cmd/plugin/cmd/init_test.go:333]: It says `cannot use provider (variable of type GeneriProviderObject) as client.Object value in argument to env.Get: generiProviderObject does not implement client.Object (ambiguous selector generiProviderObject.DeepCopyObject)`. Maybe we can keep it for now?

File: internal/controller/phases.go [hunk 7]
```diff
@@ -554,42 +552,6 @@ func (p *phaseReconciler) newClusterClient() cluster.Client {
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to parse repository url %q", providerConfig.URL())
-	}
-
-	if rURL.Scheme != httpsScheme {
-		return nil, fmt.Errorf("invalid provider url. there are no provider implementation for %q schema", rURL.Scheme)
-	}
-
-	// if the url is a GitHub repository
-	if rURL.Host == githubDomain {
```
Comment (Danil-Grigorev) [inline internal/controller/phases.go:571]: I think these constants can be removed now
```

	httpsScheme             = "https"
	githubDomain            = "github.com"
	gitlabHostPrefix        = "gitlab."
	gitlabPackagesAPIPrefix = "/api/v4/projects/"
```

**Review by Danil-Grigorev — COMMENTED** [review]
/approve

One nit, otherwise looks good

File: internal/controller/phases.go (hunk 7 above)
Comment (Fedosin) [reply internal/controller/phases.go:571]: done!

**Review by alexander-demicev — APPROVED** [review]
/lgtm

**Pull request merged**